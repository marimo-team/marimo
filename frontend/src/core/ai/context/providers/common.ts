/* Copyright 2024 Marimo. All rights reserved. */

export const Boosts = {
  LOCAL_TABLE: 5,
  REMOTE_TABLE: 4,
  VARIABLE: 3,
  ERROR: 2,
} as const;
