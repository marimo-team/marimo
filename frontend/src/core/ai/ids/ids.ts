/* Copyright 2024 Marimo. All rights reserved. */

import type { TypedString } from "@/utils/typed";

export const PROVIDERS = [
  "openai",
  "anthropic",
  "google",
  "ollama",
  "bedrock",
  "deepseek",
  "azure",
] as const;
export type ProviderId = (typeof PROVIDERS)[number];

export type ShortModelId = TypedString<"ShortModelId">;

/**
 * Qualified model id
 * `provider_id/short_model_id`;
 */
export type QualifiedModelId = `${ProviderId}/${ShortModelId}`;

export class AiModelId {
  readonly providerId: ProviderId;
  readonly shortModelId: ShortModelId;

  constructor(providerId: ProviderId, shortModelId: ShortModelId) {
    this.providerId = providerId;
    this.shortModelId = shortModelId;
  }

  get id(): QualifiedModelId {
    return `${this.providerId}/${this.shortModelId}`;
  }

  static parse(id: string): AiModelId {
    if (!id.includes("/")) {
      const providerId = guessProviderId(id);
      return new AiModelId(providerId, id as ShortModelId);
    }

    const [providerId, ...shortModelId] = id.split("/");
    return new AiModelId(
      providerId as ProviderId,
      shortModelId.join("/") as ShortModelId,
    );
  }
}

function guessProviderId(id: string): ProviderId {
  if (id.startsWith("gpt") || id.startsWith("o3") || id.startsWith("o1")) {
    return "openai";
  }
  if (id.startsWith("claude")) {
    return "anthropic";
  }
  if (id.startsWith("gemini") || id.startsWith("google")) {
    return "google";
  }
  if (id.startsWith("deepseek")) {
    return "deepseek";
  }
  return "ollama";
}

export function isKnownAIProvider(providerId: ProviderId): boolean {
  return PROVIDERS.includes(providerId);
}
