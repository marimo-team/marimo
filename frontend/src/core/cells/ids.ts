/* Copyright 2024 Marimo. All rights reserved. */
/* eslint-disable @typescript-eslint/no-redeclare */

import { invariant } from "@/utils/invariant";
import { TypedString } from "../../utils/typed";

const lowercase = "abcdefghijklmnopqrstuvwxyz";
const uppercase = lowercase.toUpperCase();
const alphabet = lowercase + uppercase;

/**
 * A typed CellId
 */
export type CellId = TypedString<"CellId">;
export const CellId = {
  /**
   * Create a new CellId, a random 4 letter string.
   */
  create(): CellId {
    let id = "";
    for (let i = 0; i < 4; i++) {
      id += alphabet[Math.floor(Math.random() * alphabet.length)];
    }
    return id as CellId;
  },
};

/**
 * The id of an HTML element that represents a cell
 */
export type HTMLCellId = `cell-${CellId}`;
export const HTMLCellId = {
  /**
   * Create a new HTMLCellId
   */
  create(cellId: CellId): HTMLCellId {
    invariant(cellId != null, "cellId is required");

    return `cell-${cellId}`;
  },
  parse(htmlCellId: HTMLCellId): CellId {
    return htmlCellId.slice(5) as CellId;
  },
  /**
   * Get the cell container ancestor of `element`, if any.
   */
  findElement(element: Element): (Element & { id: HTMLCellId }) | null {
    return element.closest('div[id^="cell-"]');
  },
};

/**
 * A typed UIElementId
 */
export type UIElementId = `${CellId}-${string}`;
export const UIElementId = {
  parse(element: Element): UIElementId | null {
    return element.getAttribute("object-id") as UIElementId | null;
  },
  parseOrThrow(element: Element): UIElementId {
    const id = UIElementId.parse(element);
    invariant(id, "<marimo-ui-element/> missing object-id attribute");
    return id;
  },
};
