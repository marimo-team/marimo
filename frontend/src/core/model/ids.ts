/* Copyright 2023 Marimo. All rights reserved. */
/* eslint-disable @typescript-eslint/no-redeclare */

import { invariant } from "@/utils/invariant";
import { TypedString } from "./typed";

/**
 * A typed CellId
 */
export type CellId = TypedString<"CellId">;
let counter = 0;
export const CellId = {
  /**
   * Create a new CellId
   */
  create(): CellId {
    const next = counter++;
    return next.toString() as CellId;
  },

  /**
   * Reset id provider.
   */
  reset() {
    counter = 0;
  },
};

/**
 * The id of an HTML element that represents a cell
 */
export type HTMLCellId = `cell-${CellId}`;
export const HTMLCellId = {
  /**
   * Create a new HTMLCellId
   */
  create(cellId: CellId): HTMLCellId {
    invariant(cellId != null, "cellId is required");

    return `cell-${cellId}` as HTMLCellId;
  },
  parse(htmlCellId: HTMLCellId): CellId {
    return htmlCellId.slice(5) as CellId;
  },
  /**
   * Get the cell container ancestor of `element`, if any.
   */
  findElement(element: Element): (Element & { id: HTMLCellId }) | null {
    return element.closest('div[id^="cell-"]');
  },
};
