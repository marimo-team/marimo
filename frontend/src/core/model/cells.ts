/* Copyright 2023 Marimo. All rights reserved. */
import { createRef } from "react";
import { OutputMessage } from "../kernel/messages";
import { SerializedEditorState } from "../codemirror/types";
import { CellHandle } from "../../editor/Cell";
import { CellId } from "./ids";

export const DEFAULT_CELL_NAME = "__";

/**
 * The status of a cell.
 *
 * queued: queued by the kernel.
 * running: currently executing.
 * idle: not running.
 * stale: stale, likely because the cell was disabled or an ancestor was disabled.
 * disabled-transitively: disabled because an ancestor was disabled.
 */
export type CellStatus =
  | "queued"
  | "running"
  | "idle"
  | "stale"
  | "disabled-transitively";

/**
 * Create a new cell with default state.
 */
export function createCell({
  key,
  ref = createRef(),
  name = DEFAULT_CELL_NAME,
  code = "",
  output = null,
  consoleOutputs = [],
  status = "idle",
  edited = false,
  interrupted = false,
  errored = false,
  stopped = false,
  runElapsedTimeMs = null,
  runStartTimestamp = null,
  lastCodeRun = null,
  serializedEditorState = null,
  config = {},
}: Partial<CellState> & Pick<CellState, "key">): CellState {
  return {
    key: key,
    ref: ref,
    config: config,
    name: name,
    output: output,
    code: code,
    status: status,
    edited: edited,
    interrupted: interrupted,
    errored: errored,
    stopped: stopped,
    runElapsedTimeMs: runElapsedTimeMs,
    runStartTimestamp: runStartTimestamp,
    lastCodeRun: lastCodeRun,
    consoleOutputs: consoleOutputs,
    serializedEditorState: serializedEditorState,
  };
}

export interface CellState {
  /** unique key */
  key: CellId;
  /** user-given name, or default */
  name: string;
  /** current contents of the editor */
  code: string;
  /** a message encoding the cell's output */
  output: OutputMessage | null;
  /** messages encoding the cell's console outputs. */
  consoleOutputs: OutputMessage[];
  /** current status of the cell */
  status: CellStatus;
  /** whether this cell has been modified since its last run */
  edited: boolean;
  /** whether this cell has been interrupted since its last run */
  interrupted: boolean;
  /** whether this cell was stopped with mo.stop */
  stopped: boolean;
  /** snapshot of code that was last run */
  lastCodeRun: string | null;
  /**
   * whether marimo encountered an error when trying to register or run
   * this cell (such as a multiple definition error)
   */
  errored: boolean;
  /** run start time, as seconds since epoch */
  runStartTimestamp: number | null;
  /** run elapsed time, in milliseconds */
  runElapsedTimeMs: number | null;
  /** serialized state of the underyling editor */
  serializedEditorState: SerializedEditorState | null;

  /** handle to access the underlying cell */
  ref: React.RefObject<CellHandle>;

  /** cell configuration */
  config: CellConfig;
}

export interface CellConfig {
  /**
   * If true, the cell and its descendants are unable to run.
   */
  disabled?: boolean;
}
