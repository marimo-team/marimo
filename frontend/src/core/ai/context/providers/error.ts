/* Copyright 2024 Marimo. All rights reserved. */

import type { Completion } from "@codemirror/autocomplete";
import { cellErrorsAtom } from "@/core/cells/cells";
import type { CellId } from "@/core/cells/ids";
import type { MarimoError } from "@/core/kernel/messages";
import type { JotaiStore } from "@/core/state/jotai";
import { PluralWord } from "@/utils/pluralize";
import { type AIContextItem, AIContextProvider } from "../registry";
import { contextToXml } from "../utils";
import { Boosts } from "./common";

export interface ErrorContextItem extends AIContextItem {
  type: "error";
  data: {
    type: "all-errors";
    errors: Array<{
      cellId: CellId;
      cellName: string;
      errorData: MarimoError[];
    }>;
  };
}

function describeError(error: MarimoError): string {
  if (error.type === "setup-refs") {
    return "The setup cell cannot have references";
  }
  if (error.type === "cycle") {
    return "This cell is in a cycle";
  }
  if (error.type === "multiple-defs") {
    return `The variable '${error.name}' was defined by another cell`;
  }
  if (error.type === "import-star") {
    return error.msg;
  }
  if (error.type === "ancestor-stopped") {
    return error.msg;
  }
  if (error.type === "ancestor-prevented") {
    return error.msg;
  }
  if (error.type === "exception") {
    return error.msg;
  }
  if (error.type === "strict-exception") {
    return error.msg;
  }
  if (error.type === "interruption") {
    return "This cell was interrupted and needs to be re-run";
  }
  if (error.type === "syntax") {
    return error.msg;
  }
  if (error.type === "unknown") {
    return error.msg;
  }
  return "Unknown error";
}

const errorsTxt = new PluralWord("error", "errors");
export class ErrorContextProvider extends AIContextProvider<ErrorContextItem> {
  readonly title = "Errors";
  readonly mentionPrefix = "@";
  readonly contextType = "error";
  private store: JotaiStore;

  constructor(store: JotaiStore) {
    super();
    this.store = store;
  }

  getItems(): ErrorContextItem[] {
    const errors = this.store.get(cellErrorsAtom);
    if (errors.length === 0) {
      return [];
    }

    return [
      {
        uri: this.asURI("all"),
        name: "Errors",
        type: this.contextType,
        data: {
          type: "all-errors",
          errors: errors.map((error) => ({
            cellId: error.cellId,
            cellName: error.cellName,
            errorData: error.output.data,
          })),
        },
        description: "All errors in the notebook",
      },
    ];

    // TODO: maybe handle single errors or grouped by types
  }

  formatCompletion(item: ErrorContextItem): Completion {
    if (item.data.type === "all-errors") {
      return {
        label: "@Errors",
        displayLabel: "Errors",
        detail: `${item.data.errors.length} ${errorsTxt.pluralize(item.data.errors.length)}`,
        boost: Boosts.ERROR,
        type: "error",
        apply: "@Errors",
        section: "Errors",
        info: () => {
          const infoContainer = document.createElement("div");
          infoContainer.classList.add(
            "mo-cm-tooltip",
            "docs-documentation",
            "min-w-[200px]",
            "flex",
            "flex-col",
            "gap-1",
            "p-2",
          );

          const headerDiv = document.createElement("div");
          headerDiv.classList.add("flex", "flex-col", "gap-1");

          const nameDiv = document.createElement("div");
          nameDiv.classList.add("font-bold", "text-base");
          nameDiv.textContent = "Errors";
          headerDiv.append(nameDiv);

          const descriptionDiv = document.createElement("div");
          descriptionDiv.classList.add("text-sm", "text-muted-foreground");
          descriptionDiv.textContent = `${item.data.errors.length} ${errorsTxt.pluralize(item.data.errors.length)}`;
          headerDiv.append(descriptionDiv);

          infoContainer.append(headerDiv);

          return infoContainer;
        },
      };
    }

    return {
      label: "Error",
      displayLabel: "Error",
      boost: Boosts.ERROR,
    };
  }

  formatContext(item: ErrorContextItem): string {
    const { data } = item;

    const xmls = data.errors.map((err) => {
      return contextToXml({
        type: this.contextType,
        data: {
          name: err.cellName || `Cell ${err.cellId}`,
          description: err.errorData
            .map((err) => describeError(err))
            .join("\n"),
        },
      });
    });

    const xml = xmls.join("\n\n");
    return xml;
  }
}
