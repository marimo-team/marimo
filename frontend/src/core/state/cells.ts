/* Copyright 2023 Marimo. All rights reserved. */
import { atom, useAtomValue, useSetAtom } from "jotai";
import { ReducerWithoutAction, useMemo } from "react";
import { CellMessage } from "../kernel/messages";
import { CellConfig, CellState, createCell } from "../model/cells";
import {
  scrollToBottom,
  scrollToTop,
  focusAndScrollCellIntoView,
} from "../model/scrollCellIntoView";
import { arrayMove } from "@dnd-kit/sortable";
import { CellId } from "../model/ids";
import { prepareCellForExecution, transitionCell } from "./cell";
import { store } from "./jotai";
import { createReducer } from "../../utils/createReducer";
import { arrayInsert, arrayDelete } from "@/utils/arrays";
import { foldAllBulk, unfoldAllBulk } from "../codemirror/editing/commands";

/* The array of cells on the page, together with a history of
 * deleted cells to implement an "undo delete" action
 */
export interface CellsAndHistory {
  /**
   * The array of cells on the page
   */
  present: CellState[];
  /**
   * Tuples of deleted cells, represented by (cell name, serialized editor
   * config, and insertion index), so that cell deletion can be undone
   *
   * (CodeMirror types the serialized config as any.)
   */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  history: Array<[string, any, number]>;

  /**
   * Key of cell to scroll to; typically set by actions that re-order the cell
   * array. Call the SCROLL_TO_TARGET action to scroll to the specified cell
   * and clear this field.
   */
  scrollKey: CellId | null;
}

function initialCellState(): CellsAndHistory {
  return {
    present: [],
    history: [],
    scrollKey: null,
  };
}

const { reducer, createActions } = createReducer(initialCellState, {
  createNewCell: (state, action: { cellId: CellId; before: boolean }) => {
    const { cellId, before } = action;
    const index = state.present.findIndex((cell) => cell.key === cellId);
    const insertionIndex = before ? index : index + 1;
    const cell = createCell({ key: CellId.create() });

    return {
      ...state,
      present: arrayInsert(state.present, insertionIndex, cell),
      scrollKey: cell.key,
    };
  },
  moveCell: (state, action: { cellId: CellId; before: boolean }) => {
    const { cellId, before } = action;
    const index = state.present.findIndex((cell) => cell.key === cellId);
    const cell = state.present[index];
    if (before && index === 0) {
      return {
        ...state,
        present: [cell, ...state.present.slice(1)],
        scrollKey: cellId,
      };
    } else if (!before && index === state.present.length - 1) {
      return {
        ...state,
        present: [...state.present.slice(0, -1), cell],
        scrollKey: cellId,
      };
    }

    return before
      ? {
          ...state,
          present: arrayMove(state.present, index, index - 1),
          scrollKey: cellId,
        }
      : {
          ...state,
          present: arrayMove(state.present, index, index + 1),
          scrollKey: cellId,
        };
  },
  dropCellOver: (state, action: { cellId: CellId; overCellId: CellId }) => {
    const { cellId, overCellId } = action;
    const fromIndex = state.present.findIndex((cell) => cell.key === cellId);
    const toIndex = state.present.findIndex((cell) => cell.key === overCellId);
    return {
      ...state,
      present: arrayMove(state.present, fromIndex, toIndex),
      scrollKey: null,
    };
  },
  focusCell: (state, action: { cellId: CellId; before: boolean }) => {
    if (state.present.length === 0) {
      return state;
    }

    const { cellId, before } = action;
    const index = state.present.findIndex((cell) => cell.key === cellId);
    let focusIndex = before ? index - 1 : index + 1;
    // clamp
    focusIndex = Math.max(0, Math.min(focusIndex, state.present.length - 1));
    // can scroll immediately, without setting scrollKey in state, because
    // CellArray won't need to re-render
    focusAndScrollCellIntoView(state.present[focusIndex]);
    return state;
  },
  focusTopCell: (state) => {
    if (state.present.length === 0) {
      return state;
    }

    state.present[0].ref.current?.editorView.focus();
    scrollToTop();
    return state;
  },
  focusBottomCell: (state) => {
    if (state.present.length === 0) {
      return state;
    }

    state.present[state.present.length - 1].ref.current?.editorView.focus();
    scrollToBottom();
    return state;
  },
  sendToTop: (state, action: { cellId: CellId }) => {
    if (state.present.length === 0) {
      return state;
    }

    const { cellId } = action;
    const index = state.present.findIndex((cell) => cell.key === cellId);
    return {
      ...state,
      present: arrayMove(state.present, index, 0),
      scrollKey: cellId,
    };
  },
  sendToBottom: (state, action: { cellId: CellId }) => {
    if (state.present.length === 0) {
      return state;
    }

    const { cellId } = action;
    const index = state.present.findIndex((cell) => cell.key === cellId);
    return {
      ...state,
      present: arrayMove(state.present, index, state.present.length - 1),
      scrollKey: cellId,
    };
  },
  deleteCell: (state, action: { cellId: CellId }) => {
    const cellId = action.cellId;
    if (state.present.length === 1) {
      return state;
    }

    const index = state.present.findIndex((cell) => cell.key === cellId);
    const focusIndex = index === 0 ? 1 : index - 1;
    const scrollKey = state.present[focusIndex].key;

    return {
      present: arrayDelete(state.present, index),
      history: [
        ...state.history,
        [
          state.present[index].name,
          state.present[index].ref.current?.editorStateJSON(),
          index,
        ],
      ],
      scrollKey: scrollKey,
    };
  },
  undoDeleteCell: (state) => {
    if (state.history.length === 0) {
      return state;
    } else {
      const mostRecentlyDeleted = state.history[state.history.length - 1];

      const name = mostRecentlyDeleted[0];
      const serializedEditorState = mostRecentlyDeleted[1] || {
        doc: "",
      };
      const index = mostRecentlyDeleted[2];
      const undoCell = createCell({
        key: CellId.create(),
        name,
        initialContents: serializedEditorState.doc,
        code: serializedEditorState.doc,
        edited: serializedEditorState.doc.trim().length > 0,
        serializedEditorState,
      });
      return {
        ...state,
        present: arrayInsert(state.present, index, undoCell),
        history: state.history.slice(0, -1),
      };
    }
  },
  updateCellCode: (
    state,
    action: {
      cellId: CellId;
      code: string;
      /**
       * Whether or not the update is a formatting change,
       * if so, the 'edited' state will be handled differently.
       */
      formattingChange: boolean;
    }
  ) => {
    const { cellId, code, formattingChange } = action;
    const cellToUpdate = state.present.find((cell) => cell.key === cellId);

    if (!cellToUpdate || cellToUpdate.code === code) {
      return state;
    }

    return updateCell(state, cellId, (cell) => {
      // Formatting-only change means we can re-use the last code run
      // if it was not previously edited. And we don't change the edited state.
      return formattingChange
        ? {
            ...cell,
            code: code,
            lastCodeRun: cell.edited ? cell.lastCodeRun : code,
          }
        : {
            ...cell,
            code: code,
            edited: code.trim() !== cell.lastCodeRun,
          };
    });
  },
  updateCellConfig: (
    state,
    action: { cellId: CellId; config: Partial<CellConfig> }
  ) => {
    const { cellId, config } = action;
    return updateCell(state, cellId, (cell) => {
      return {
        ...cell,
        config: { ...cell.config, ...config },
      };
    });
  },
  prepareForRun: (state, action: { cellId: CellId }) => {
    const cellToUpdate = state.present.find(
      (cell) => cell.key === action.cellId
    );
    if (!cellToUpdate) {
      return state;
    }
    return updateCell(state, action.cellId, (cell) => {
      return prepareCellForExecution(cell);
    });
  },
  handleCellMessage: (
    state,
    action: { cellId: CellId; message: CellMessage }
  ) => {
    const { cellId, message } = action;
    return updateCell(state, cellId, (cell) => {
      return transitionCell(cell, message);
    });
  },
  setCells: (state, cells: CellState[]) => {
    return {
      ...state,
      present: cells,
    };
  },
  /**
   * Move focus to next cell
   *
   * Creates a new cell if the current cell is the last one in the array.
   *
   * If needed, scrolls newly created or focused cell into view.
   *
   * Replicates Shift+Enter functionality of Jupyter
   */
  moveToNextCell: (state, action: { cellId: CellId; before: boolean }) => {
    const { cellId, before } = action;
    const index = state.present.findIndex((cell) => cell.key === cellId);
    const nextCellIndex = before ? index - 1 : index + 1;
    // Create a new cell at the end; no need to update scrollKey,
    // because cell will be created with autoScrollIntoView
    if (nextCellIndex === state.present.length) {
      const newCell = createCell({
        key: CellId.create(),
      });
      return {
        ...state,
        present: [...state.present, newCell],
      };
      // Create a new cell at the beginning; again, no need to update
      // scrollKey
    } else if (nextCellIndex === -1) {
      const newCell = createCell({
        key: CellId.create(),
      });
      return {
        ...state,
        present: [newCell, ...state.present],
      };
    } else {
      // Just focus, no state change
      focusAndScrollCellIntoView(state.present[nextCellIndex]);
      return state;
    }
  },
  scrollToTarget: (state) => {
    // Scroll to the specified cell and clear the scroll key.
    const scrollKey = state.scrollKey;
    if (scrollKey === null) {
      return state;
    } else {
      const index = state.present.findIndex((cell) => cell.key === scrollKey);

      // Special-case scrolling to the end of the page: bug in Chrome where
      // browser fails to scrollIntoView an element at the end of a long page
      if (index === state.present.length - 1) {
        state.present[state.present.length - 1].ref.current?.editorView.focus();
        scrollToBottom();
      } else {
        focusAndScrollCellIntoView(state.present[index]);
      }

      return {
        ...state,
        scrollKey: null,
      };
    }
  },
  foldAll: (state) => {
    const targets = state.present.map((cell) => cell.ref.current?.editorView);
    foldAllBulk(targets);
    return state;
  },
  unfoldAll: (state) => {
    const targets = state.present.map((cell) => cell.ref.current?.editorView);
    unfoldAllBulk(targets);
    return state;
  },
});

// Helper function to update a cell in the array
function updateCell(
  state: CellsAndHistory,
  cellId: CellId,
  cellReducer: ReducerWithoutAction<CellState>
) {
  return {
    ...state,
    present: state.present.map((cell) =>
      cell.key === cellId ? cellReducer(cell) : cell
    ),
  };
}

const cellsAtom = atom<CellsAndHistory>(initialCellState());

/**
 * Get the array of cells.
 */
export const useCells = () => useAtomValue(cellsAtom);

/**
 * Get the editor views for all cells.
 */
export const getAllEditorViews = () => {
  const cells = store.get(cellsAtom).present;
  return cells
    .map((cell) => cell.ref.current?.editorView)
    .flatMap((x) => (x ? [x] : []));
};

/**
 * Use this hook to dispatch cell actions. This hook will not cause a re-render
 * when cells change.
 */
export function useCellActions() {
  const setState = useSetAtom(cellsAtom);

  return useMemo(() => {
    const actions = createActions((action) => {
      setState((state) => reducer(state, action));
    });
    return actions;
  }, [setState]);
}

export type CellActions = ReturnType<typeof createActions>;

/**
 * This is exported for testing purposes only.
 */
export const exportedForTesting = {
  reducer,
  createActions,
  initialCellState,
};
