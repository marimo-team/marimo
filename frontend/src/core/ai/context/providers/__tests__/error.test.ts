/* Copyright 2024 Marimo. All rights reserved. */
/* eslint-disable @typescript-eslint/no-explicit-any */

import { createStore } from "jotai";
import { beforeEach, describe, expect, it } from "vitest";
import { MockNotebook } from "@/__mocks__/notebook";
import { notebookAtom } from "@/core/cells/cells";
import { type CellId, CellId as CellIdClass } from "@/core/cells/ids";
import { Boosts } from "../common";
import { ErrorContextProvider } from "../error";

describe("ErrorContextProvider", () => {
  let provider: ErrorContextProvider;
  let store: ReturnType<typeof createStore>;

  beforeEach(() => {
    store = createStore();
    provider = new ErrorContextProvider(store);
  });

  const createMockNotebookWithErrors = (
    errors: Parameters<typeof MockNotebook.notebookStateWithErrors>[0],
  ) => {
    const notebookState = MockNotebook.notebookStateWithErrors(errors);
    store.set(notebookAtom, notebookState);
  };

  describe("provider properties", () => {
    it("should have correct provider properties", () => {
      expect(provider.title).toBe("Errors");
      expect(provider.mentionPrefix).toBe("@");
      expect(provider.contextType).toBe("error");
    });
  });

  describe("getItems", () => {
    it("should return empty array when no errors", () => {
      const items = provider.getItems();
      expect(items).toEqual([]);
    });

    it("should return error items when errors exist", () => {
      const cellId1 = CellIdClass.create();
      const cellId2 = CellIdClass.create();

      createMockNotebookWithErrors([
        {
          cellId: cellId1,
          cellName: "Cell 1",
          errorData: [MockNotebook.errors.syntax("Invalid syntax")],
        },
        {
          cellId: cellId2,
          cellName: "Cell 2",
          errorData: [MockNotebook.errors.exception("Runtime error")],
        },
      ]);

      const items = provider.getItems();

      expect(items).toHaveLength(1);
      expect(items[0]).toMatchObject({
        name: "Errors",
        type: "error",
        description: "All errors in the notebook",
        data: {
          type: "all-errors",
          errors: [
            {
              cellId: cellId1,
              cellName: "Cell 1",
              errorData: [{ type: "syntax", msg: "Invalid syntax" }],
            },
            {
              cellId: cellId2,
              cellName: "Cell 2",
              errorData: [{ type: "exception", msg: "Runtime error" }],
            },
          ],
        },
      });
    });

    it("should handle cells without names", () => {
      const cellId = CellIdClass.create();

      createMockNotebookWithErrors([
        {
          cellId,
          cellName: "",
          errorData: [MockNotebook.errors.syntax("Invalid syntax")],
        },
      ]);

      const items = provider.getItems();
      expect(items[0].data.errors[0].cellName).toBe("");
    });
  });

  describe("formatCompletion", () => {
    it("should format completion for all-errors type", () => {
      const cellId1 = CellIdClass.create();
      const cellId2 = CellIdClass.create();

      createMockNotebookWithErrors([
        {
          cellId: cellId1,
          cellName: "Cell 1",
          errorData: [MockNotebook.errors.syntax("Invalid syntax")],
        },
        {
          cellId: cellId2,
          cellName: "Cell 2",
          errorData: [MockNotebook.errors.exception("Runtime error")],
        },
      ]);

      const items = provider.getItems();
      const completion = provider.formatCompletion(items[0]);

      expect(completion).toMatchObject({
        label: "@Errors",
        displayLabel: "Errors",
        detail: "2 errors",
        boost: Boosts.ERROR,
        type: "error",
        apply: "@Errors",
        section: "Errors",
      });

      // Test the info function
      expect(completion.info).toBeDefined();
      if (typeof completion.info === "function") {
        const infoResult = completion.info(completion);
        if (
          infoResult &&
          typeof infoResult === "object" &&
          "dom" in infoResult
        ) {
          const infoElement = infoResult.dom as HTMLElement;
          expect(infoElement.tagName).toBe("DIV");
          expect(infoElement.textContent).toContain("Errors");
          expect(infoElement.textContent).toContain("2 errors");
        } else if (infoResult && "tagName" in (infoResult as any)) {
          const infoElement = infoResult as HTMLElement;
          expect(infoElement.tagName).toBe("DIV");
          expect(infoElement.textContent).toContain("Errors");
          expect(infoElement.textContent).toContain("2 errors");
        }
      }
    });

    it("should handle single error correctly", () => {
      const cellId = CellIdClass.create();

      createMockNotebookWithErrors([
        {
          cellId,
          cellName: "Cell 1",
          errorData: [MockNotebook.errors.syntax("Invalid syntax")],
        },
      ]);

      const items = provider.getItems();
      const completion = provider.formatCompletion(items[0]);
      expect(completion.detail).toBe("1 error");
    });

    it("should handle fallback for unknown error types", () => {
      const item = {
        uri: "error://unknown",
        name: "Unknown Error",
        type: "error" as const,
        data: {
          type: "unknown" as any,
          errors: [],
        },
        description: "Unknown error type",
      };

      const completion = provider.formatCompletion(item);
      expect(completion).toMatchObject({
        label: "Error",
        displayLabel: "Error",
        boost: Boosts.ERROR,
      });
    });
  });

  describe("formatContext", () => {
    it("should format context for basic errors", () => {
      const cellId = CellIdClass.create();

      createMockNotebookWithErrors([
        {
          cellId,
          cellName: "Cell 1",
          errorData: [MockNotebook.errors.syntax("Invalid syntax")],
        },
      ]);

      const items = provider.getItems();
      const context = provider.formatContext(items[0]);
      expect(context).toMatchSnapshot("basic-error-context");
    });

    it("should format context for multiple error types", () => {
      const cellId1 = "cell-1" as CellId;
      const cellId2 = "cell-2" as CellId;

      createMockNotebookWithErrors([
        {
          cellId: cellId1,
          cellName: "Cell 1",
          errorData: [
            MockNotebook.errors.syntax("Invalid syntax"),
            MockNotebook.errors.exception("Runtime error"),
          ],
        },
        {
          cellId: cellId2,
          cellName: "",
          errorData: [
            MockNotebook.errors.cycle(),
            MockNotebook.errors.multipleDefs("variable_x"),
          ],
        },
      ]);

      const items = provider.getItems();
      const context = provider.formatContext(items[0]);

      // Check for expected content instead of exact snapshot match due to random cell IDs
      expect(context).toMatchInlineSnapshot(`
        "<error name="Cell 1" description="Invalid syntax
        Runtime error"></error>

        <error name="Cell cell-2" description="This cell is in a cycle
        The variable 'variable_x' was defined by another cell"></error>"
      `);
    });

    it("should handle cells without names", () => {
      const cellId = CellIdClass.create();

      createMockNotebookWithErrors([
        {
          cellId,
          cellName: "",
          errorData: [MockNotebook.errors.syntax("Invalid syntax")],
        },
      ]);

      const items = provider.getItems();
      const context = provider.formatContext(items[0]);
      expect(context).toContain(`Cell ${cellId}`);
    });
  });

  describe("error description handling", () => {
    it("should handle all error types correctly", () => {
      // Note: ancestor-* error types are intentionally filtered out by cellErrorsAtom
      const errorTypes = [
        {
          error: MockNotebook.errors.setupRefs(),
          expected: "The setup cell cannot have references",
        },
        {
          error: MockNotebook.errors.cycle(),
          expected: "This cell is in a cycle",
        },
        {
          error: MockNotebook.errors.multipleDefs("var_x"),
          expected: "The variable 'var_x' was defined by another cell",
        },
        {
          error: MockNotebook.errors.importStar("Import star error"),
          expected: "Import star error",
        },
        {
          error: MockNotebook.errors.exception("Runtime exception"),
          expected: "Runtime exception",
        },
        {
          error: MockNotebook.errors.strictException("Strict exception", "ref"),
          expected: "Strict exception",
        },
        {
          error: MockNotebook.errors.interruption(),
          expected: "This cell was interrupted and needs to be re-run",
        },
        {
          error: MockNotebook.errors.syntax("Syntax error"),
          expected: "Syntax error",
        },
        {
          error: MockNotebook.errors.unknown("Unknown error"),
          expected: "Unknown error",
        },
      ];

      for (const { error, expected } of errorTypes) {
        // Create a fresh store and provider for each error type
        const testStore = createStore();
        const testProvider = new ErrorContextProvider(testStore);

        const { notebookState } = MockNotebook.cellWithErrors("Cell 1", [
          error,
        ]);
        testStore.set(notebookAtom, notebookState);

        const items = testProvider.getItems();
        expect(items).toHaveLength(1);
        const context = testProvider.formatContext(items[0]);
        expect(context).toContain(expected);
      }
    });

    it("should handle unknown error type gracefully", () => {
      const error = { type: "unknown-type" as any };
      const { notebookState } = MockNotebook.cellWithErrors("Cell 1", [error]);
      store.set(notebookAtom, notebookState);

      const items = provider.getItems();
      const context = provider.formatContext(items[0]);
      expect(context).toContain("Unknown error");
    });
  });
});
