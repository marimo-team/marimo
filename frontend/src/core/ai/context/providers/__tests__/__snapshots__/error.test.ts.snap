// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ErrorContextProvider > formatContext > should format context for basic errors > basic-error-context 1`] = `"<error name="Cell 1" description="Invalid syntax"></error>"`;
